core/src/main/java/bio/terra/pearl/core/service/DataAuditedService.java

```diff
@@ -0,0 +1,164 @@
+package bio.terra.pearl.core.service;
+
+import bio.terra.pearl.core.dao.BaseMutableJdbiDao;
+import bio.terra.pearl.core.model.BaseEntity;
+import bio.terra.pearl.core.model.workflow.DataAuditInfo;
+import bio.terra.pearl.core.model.workflow.DataChangeRecord;
+import bio.terra.pearl.core.service.workflow.DataChangeRecordService;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * base class for entities that should have changes tracked via DataChangeRecords.
+ * This currently allows 'auditInfo' to be null--a later PR should remove that, once we upgrade all the calling services
+ * to also support passing auditInfo to create/deletes
+ * */
+public abstract class DataAuditedService<M extends BaseEntity, D extends BaseMutableJdbiDao<M>> {
```

Comment:
Is the idea that this class will be implicitly tested by higher level tests, or is there not a current way to deal with an actual DB for automated tests (thus testing at this level would not be very effective)?

Reply:
Yes, the idea is that this is tested by child classes.  All our automated tests are run against an actual postgres db (that's in the same docker container as the application postgres db--see `core/src/test/resources/application.yaml` for the config).  Once we get a sense of how many services will inherit from this, it might be worth testing this independently with some model subclass, but for now the implicit coverage by the one-and-only subclass should be good enough